### PR TITLE
Reduce the size of the indices we keep for many global search operations

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @sharwell   A way to make our index smaller.  It stores less strings in the index, at the cost of having to go back to a syntax tree when it finds a hit later on.  

Note that this is still better than Def15, where we'd have to go back to a *symbol* when we found a match.  But it's not ideal as it would be nice if just finding a match provided enough information to supply the UI with all the details it needed.

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs [hunk 1]
```diff
@@ -75,6 +78,7 @@ internal abstract partial class AbstractNavigateToSearchService
 
                 cancellationToken.ThrowIfCancellationRequested();
                 var declarationInfo = await document.GetSyntaxTreeIndexAsync(cancellationToken).ConfigureAwait(false);
+                var root = default(SyntaxNode);
```
Comment (Pilchie) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs]: I hate this.  Just use `SyntaxNode root = null;`  If you do this, I have to think about whether `SyntaxNode` is a struct or not.

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs [hunk 2]
```diff
@@ -772,6 +772,77 @@ public bool IsTopLevelNodeWithMembers(SyntaxNode node)
...
         }
 
+        public void GetSuffixAndContainerDisplayName(SyntaxNode node, out string nameSuffix, out string containerDisplayName)
+        {
+            switch (node.Kind())
+            {
+                case SyntaxKind.ClassDeclaration:
+                    var classDecl = (ClassDeclarationSyntax)node;
+                    nameSuffix = GetTypeParameterSuffix(classDecl.TypeParameterList);
+                    containerDisplayName = GetContainerDisplayName(node.Parent);
```
Comment (Pilchie) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs]: Every single one of these does this - can it just be done before the switch?

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs [hunk 3]
```diff
@@ -772,6 +772,77 @@ public bool IsTopLevelNodeWithMembers(SyntaxNode node)
...
+                    nameSuffix = GetConstructorSuffix(ctorDecl);
+                    containerDisplayName = GetContainerDisplayName(node.Parent);
+                    return;
+                case SyntaxKind.DelegateDeclaration:
+                    var delegateDecl = (DelegateDeclarationSyntax)node;
+                    nameSuffix = GetTypeParameterSuffix(delegateDecl.TypeParameterList);
+                    containerDisplayName = GetContainerDisplayName(node.Parent);
+                    return;
+                case SyntaxKind.EnumDeclaration:
+                    var enumDecl = (EnumDeclarationSyntax)node;
```
Comment (Pilchie) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs]: So many of these are unused?

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs [hunk 4]
```diff
@@ -772,6 +772,77 @@ public bool IsTopLevelNodeWithMembers(SyntaxNode node)
...
+                    return;
+                case SyntaxKind.VariableDeclarator:
+                    // could either be part of a field declaration or an event field declaration
+                    var fieldDeclaration = (BaseFieldDeclarationSyntax)node.Parent.Parent;
+                    nameSuffix = "";
+                    containerDisplayName = GetContainerDisplayName(fieldDeclaration.Parent);
+                    return;
+            }
+
+            nameSuffix = "";
```
Comment (Pilchie) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs]: Move into a `default` with a `return` so no case can forget to initialize and return?

File: src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxFactsService.vb [hunk 5]
```diff
@@ -845,14 +845,82 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                    TypeOf node Is EnumBlockSyntax
         End Function
 
+        Public Sub GetSuffixAndContainerDisplayName(node As SyntaxNode, ByRef nameSuffix As String, ByRef containerDisplayName As String) Implements ISyntaxFactsService.GetSuffixAndContainerDisplayName
+            Select Case node.Kind()
```
Comment (Pilchie) [inline src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxFactsService.vb]: Similar comments as on the C# side.

**Comment by sharwell** [comment]
@Pilchie The goal of this change is to reduce the observed "fixed" overhead when working in Visual Studio - that is, code where allocation rate is not particularly high but caching increases the amount of long-lived memory in use. The primary beneficiary of improvements in this area is users working in memory-constrained environments. For example, while users working on Roslyn typically expect an OOM when Visual Studio runs near 4G memory in use, I found this opportunity while investigating an OOM crash that occurred while VS was using less than 1G memory.

**Comment by CyrusNajmabadi** [comment]
This was more of an experiment to see what was possible.  I'm actually feeling like i don't want to take this.  Specifically because we lower memory, but increase CPU during searching.  I'd rather us use more memory and have fast search.

That said, i think we have areas where we can still improve.  It's quite likely that lots of these DeclaredSymbolInfo structs point at identical strings (i.e. the ContainerDisplayName for all the types in a single namespace).  We can try to come up with some sort of system that allows these strings to be interned/shared to help get the majority of hte benefit here, without affecting search speeds.

**Comment by sharwell** [comment]
> We can try to come up with some sort of system that allows these strings to be interned/shared to help get the majority of the benefit here, without affecting search speeds.

I agree this would be preferable.

**Comment by CyrusNajmabadi** [comment]
Closing this out.  Was an interesting thing to try out.  But i don't think it's where we'd like to go.

**Pull request closed without merging**